File: glTest/src/problems/Problem.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package problems;

import com.jogamp.opengl.GL3;
import framework.GLApi;
import solutions.Solution;

/**
 *
 * @author GBarbieri
 */
public abstract class Problem {

    protected Solution activeSolution;

    public boolean init(GL3 gl3) {
        return false;
    }

    public void render(GL3 gl3) {

    }

    public void shutdown(GL3 gl3) {

    }

    public void clear(GL3 gl3) {

    }

    public String getName() {
        return "";
    }

    public boolean setSolution(GL3 gl3, Solution solution) {

        assert (solution == null || solution.getProblemName().equals(getName()));

        if (activeSolution != null) {
            System.out.println("Solution " + activeSolution.getName() + " - shutdown beginning.");
            activeSolution.shutdown(gl3);
            System.out.println("Solution " + activeSolution.getName() + " - shutdown complete.");
        }

        activeSolution = solution;

        // The parameters to be handed off by Init are specific to the problem being solved, so is called after this
        // by the derived class.
        return true;
    }

}

File: src/main/kotlin/main/glTest/problems/Problem.kt
package main.glTest.problems

import glm_.vec4.Vec4
import main.glTest.solutions.Solution
import org.lwjgl.opengl.GL11C.glGetError

/**
 *
 * @author GBarbieri
 */
abstract class Problem {

    open fun destroy() {}

    /** @return outDepth */
    abstract fun getClearValues(outCol: Vec4): Float

    abstract fun init(): Boolean
    abstract fun render()
    open fun shutdown() {
        setSolution(null)
    }

    abstract val name: String

    var activeSolution: Solution? = null

    open fun setSolution(solution: Solution?): Boolean {

        assert(solution == null || solution.problemName == name)

        activeSolution?.let {
            println("Solution ${it.name} - shutdown beginning.")
            it.shutdown()
            glGetError()
            println("Solution ${it.name} shutdown complete.")
        }

        activeSolution = solution

        // The parameters to be handed off by Init are specific to the problem being solved,
        // by the derived class.
        return true
    }
}
